### PR TITLE
[tune] Add Ax to Tune

### PR DESCRIPTION
**Pull request by adizim**
image

Results from 50 sample experiment

Required dependencies (working from fresh conda env w/ python3.6):

1. pip install ray
2. conda install pytorch torchvision -c pytorch
3. pip install ax-platform
4. pip install sqlalchemy

### CODE REVIEW

**Comment by adizim** [comment]
@AmplabJenkins retest this please

File: doc/source/tune-searchalg.rst [hunk 1]
```diff
@@ -136,7 +137,7 @@ Scikit-Optimize Search
 
 The ``SkOptSearch`` is a SearchAlgorithm that is backed by `Scikit-Optimize <https://scikit-optimize.github.io>`__ to perform sequential model-based hyperparameter optimization. Note that this class does not extend ``ray.tune.suggest.BasicVariantGenerator``, so you will not be able to use Tune's default variant generation/search space declaration when using SkOptSearch.
 
-In order to use this search algorithm, you will need to install Scikit-Optimize via the following command:
```
Comment (richardliaw) [inline doc/source/tune-searchalg.rst]: wait we don't want to change this right :)

File: doc/source/tune-searchalg.rst [hunk 2]
```diff
@@ -155,6 +156,30 @@ An example of this can be found in `skopt_example.py <https://github.com/ray-pro
     :show-inheritance:
     :noindex:
 
+Ax Search
+---------
+
+The ``AxSearch`` is a SearchAlgorithm that is backed by `Ax <https://ax.dev/>`__ to perform sequential model-based hyperparameter optimization. Note that this class does not extend ``ray.tune.suggest.BasicVariantGenerator``, so you will not be able to use Tune's default variant generation/search space declaration when using SkOptSearch.
```
Comment (richardliaw) [inline doc/source/tune-searchalg.rst]: maybe also include a one-line comment about what is Ax. `Ax is a  platform for understanding, managing, deploying, and automating adaptive experiments. Ax provides an easy to use interface with BoTorch, a flexible, modern library for Bayesian optimization in PyTorch.`

File: doc/source/tune-searchalg.rst [hunk 3]
```diff
@@ -155,6 +156,30 @@ An example of this can be found in `skopt_example.py <https://github.com/ray-pro
...
+
+In order to use this search algorithm, you will need to install Ax via the following commands:
+
+.. code-block:: bash
+
+    $ conda install pytorch torchvision -c pytorch  # OSX only
```
Comment (richardliaw) [inline doc/source/tune-searchalg.rst]: I would consider just linking to the PyTorch Installation page here.

File: doc/source/tune-searchalg.rst [hunk 4]
```diff
@@ -155,6 +156,30 @@ An example of this can be found in `skopt_example.py <https://github.com/ray-pro
...
+    $ pip3 install ax-platform  # all systems
+
+This algorithm requires specifying a search space and objective. You can use AxSearch like follows:
```
Comment (richardliaw) [inline doc/source/tune-searchalg.rst]:
```suggestion
This algorithm requires specifying a search space and objective. You can use `AxSearch` like follows:
```

File: python/ray/tune/examples/ax_example.py [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import ray
+from ray.tune import run
+from ray.tune.schedulers import AsyncHyperBandScheduler
+from ray.tune.suggest.ax import AxSearch
+
+import numpy as np
```
Comment (richardliaw) [inline python/ray/tune/examples/ax_example.py]: the general style is to move this before the ray imports.

File: python/ray/tune/suggest/ax.py [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+        parameters (list[dict]): Parameters in the experiment search space.
+            Required elements in the dictionaries are: "name" (name of
+            this parameter, string), "type" (type of the parameter: "range",
+            "fixed", or "choice", string), "bounds" for range parameters
+            (list of two values, lower bound first), "values" for choice
+            parameters (list of values), and "value" for fixed parameters
+            (single value).
+        objective_name (str): Name of the metric used as objective in this
+            experiment. This metric must be present in `raw_data` argument
+            to `log_data`.
```
Comment (richardliaw) [inline python/ray/tune/suggest/ax.py]: This metric must also be present in the dict reported/returned by the Trainable.

File: python/ray/tune/suggest/ax.py [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+        max_concurrent (int): Number of maximum concurrent trials. Defaults
+            to 10.
+        minimize (bool): Whether this experiment represents a minimization
```
Comment (richardliaw) [inline python/ray/tune/suggest/ax.py]: Defaults to False.

File: python/ray/tune/suggest/ax.py [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+            problem.
```
Comment (richardliaw) [inline python/ray/tune/suggest/ax.py]:
```suggestion
            problem. Defaults to False.
```

File: python/ray/tune/suggest/ax.py [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+        parameter_constraints (list[str]): Parameter constraints, such as
+            "x3 >= x4" or "x3 + x4 >= 2".
+        outcome_constraints (list[str]): Outcome constraints of form
+            "metric_name >= bound", like "m1 <= 3."
+        outcome_names (list[str]): Names of outcome constraints, must
+            equal number of outcome_constraints
```
Comment (richardliaw) [inline python/ray/tune/suggest/ax.py]:
```suggestion
            equal number of outcome_constraints.
```

File: python/ray/tune/suggest/ax.py [hunk 10]
```diff
@@ -0,0 +1,113 @@
...
+
+    def on_trial_result(self, trial_id, result):
+        pass
+
+    def on_trial_complete(self,
+                          trial_id,
+                          result=None,
+                          error=False,
+                          early_terminated=False):
+        """Pass data back to Ax
```
Comment (richardliaw) [inline python/ray/tune/suggest/ax.py]:
```suggestion
        """Pass data back to Ax.
```

File: doc/source/tune-searchalg.rst [hunk 11]
```diff
@@ -155,6 +156,30 @@ An example of this can be found in `skopt_example.py <https://github.com/ray-pro
...
+---------
+
+The ``AxSearch`` is a SearchAlgorithm that is backed by `Ax <https://ax.dev/>`__ to perform sequential model-based hyperparameter optimization. Note that this class does not extend ``ray.tune.suggest.BasicVariantGenerator``, so you will not be able to use Tune's default variant generation/search space declaration when using SkOptSearch.
+
+In order to use this search algorithm, you will need to install Ax via the following commands:
+
+.. code-block:: bash
+
+    $ conda install pytorch torchvision -c pytorch  # OSX only
+    $ pip3 install ax-platform  # all systems
```
Comment (richardliaw) [inline doc/source/tune-searchalg.rst]:
```suggestion
    $ pip install ax-platform  # all systems
```

File: python/ray/tune/suggest/ax.py [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+        >>> algo = AxSearch(parameters=parameters,
+        >>>     objective_name="hartmann6", max_concurrent=4)
+    """
+
+    def __init__(self,
+                 parameters,
+                 objective_name,
+                 max_concurrent=10,
+                 minimize=False,
+                 parameter_constraints=[],
```
Comment (richardliaw) [inline python/ray/tune/suggest/ax.py]: we should change this to `parameter_constraints=None` (along with any default kwarg that is `[]` or `{}`. There's a weird behavior with these lists being mutable and resulting in some hidden side effects if there are multiple instantiations.

For these defaults, the assignment should be like:
```
self._ax.create_experiment(
            name="ax",
            parameters=parameters,
            objective_name=objective_name,
            minimize=minimize,
            parameter_constraints=parameter_constraints or [],
            outcome_constraints=outcome_constraints or [],
...)
```

File: doc/source/tune-searchalg.rst (hunk 2 above)
Comment (richardliaw) [reply doc/source/tune-searchalg.rst]: Also says SkOptSearch here.

**Review by richardliaw — CHANGES_REQUESTED** [review]
Left a few comments (be sure to request a review again when addressed).

File: doc/source/tune-searchalg.rst [hunk 13]
```diff
@@ -155,6 +156,29 @@ An example of this can be found in `skopt_example.py <https://github.com/ray-pro
...
+Ax Search
+---------
+
+The ``AxSearch`` is a SearchAlgorithm that is backed by `Ax <https://ax.dev/>`__ to perform sequential model-based hyperparameter optimization. Ax is a platform for understanding, managing, deploying, and automating adaptive experiments. Ax provides an easy to use interface with BoTorch, a flexible, modern library for Bayesian optimization in PyTorch. Note that this class does not extend ``ray.tune.suggest.BasicVariantGenerator``, so you will not be able to use Tune's default variant generation/search space declaration when using AxSearch.
+
+In order to use this search algorithm, you will need to install PyTorch and Ax. Instructions to install PyTorch locally can be found `here <https://pytorch.org/get-started/locally/>`__. You can install Ax via the following command:
+
+.. code-block:: bash
+
+    $ pip install ax-platform
```
Comment (richardliaw) [inline doc/source/tune-searchalg.rst]:
```suggestion
    $ pip install ax-platform sqlalchemy
```

File: doc/source/tune-searchalg.rst [hunk 14]
```diff
@@ -155,6 +156,29 @@ An example of this can be found in `skopt_example.py <https://github.com/ray-pro
     :show-inheritance:
     :noindex:
 
+Ax Search
+---------
+
+The ``AxSearch`` is a SearchAlgorithm that is backed by `Ax <https://ax.dev/>`__ to perform sequential model-based hyperparameter optimization. Ax is a platform for understanding, managing, deploying, and automating adaptive experiments. Ax provides an easy to use interface with BoTorch, a flexible, modern library for Bayesian optimization in PyTorch. Note that this class does not extend ``ray.tune.suggest.BasicVariantGenerator``, so you will not be able to use Tune's default variant generation/search space declaration when using AxSearch.
+
+In order to use this search algorithm, you will need to install PyTorch and Ax. Instructions to install PyTorch locally can be found `here <https://pytorch.org/get-started/locally/>`__. You can install Ax via the following command:
```
Comment (richardliaw) [inline doc/source/tune-searchalg.rst]:
```suggestion
In order to use this search algorithm, you will need to install PyTorch, Ax, and sqlalchemy. Instructions to install PyTorch locally can be found `here <https://pytorch.org/get-started/locally/>`__. You can install Ax and sqlalchemy via the following command:
```

**Review by richardliaw — APPROVED** [review]
Pending last installation comments and this is good to go!

**Comment by richardliaw** [comment]
Thanks @adizim!

**Pull request merged**